### PR TITLE
Retire stale changewaves

### PR DESCRIPTION
**Pull request by rainersigwald**
After discussion in #9543 we decided to delay changewave retirement to a release corresponding with 9.0.100. The time is now.

### CODE REVIEW

**Comment by rainersigwald** [comment]
Pinging folks relevant to the changes that are being made permanent: @ar-may, @JanKrivanek, @Forgind, @JaynieBai.

Also @marcpopMSFT as a heads up.

File: documentation/wiki/ChangeWaves.md [hunk 1]
```diff
@@ -57,13 +57,6 @@ A wave of features is set to "rotate out" (i.e. become standard functionality) t
 - [AnyHaveMetadataValue returns false when passed an empty list](https://github.com/dotnet/msbuild/pull/8603)
 - [Log item self-expansion](https://github.com/dotnet/msbuild/pull/8581)
 
-### 17.4
```
Comment (maridematte) [inline documentation/wiki/ChangeWaves.md:60]: We are also retiring 17.6 and 17.8 right? These are still included in the active changewaves here.

Comment (rainersigwald) [reply documentation/wiki/ChangeWaves.md:60]: Thanks!

**Review by Forgind — COMMENTED** [review]
I don't think this is actually necessary, but the example in ChangeWaves-Dev.md uses Wave17_4; we could update that to 17_12 or something.

Also, I thought we were planning to always have three change waves active at a time? If we're changing that, perhaps we should update the .md to say that 🙂

Finally, I have no idea why, but apparently this is a thing?
https://github.com/dotnet/msbuild/blob/a49bf6cf54854d4afb9d6b72c9d4b2d4efc9fd30/src/Build/Resources/Strings.resx#L1301

File: src/Build/BackEnd/Components/SdkResolution/SdkResolverService.cs [hunk 2]
```diff
@@ -118,7 +110,7 @@ public virtual void ClearCaches()
         /// <inheritdoc cref="ISdkResolverService.ResolveSdk"/>
         public virtual SdkResult ResolveSdk(int submissionId, SdkReference sdk, LoggingContext loggingContext, ElementLocation sdkReferenceLocation, string solutionPath, string projectPath, bool interactive, bool isRunningInVisualStudio, bool failOnUnresolvedSdk)
         {
-            // If we are running in .NET core, we ask the built-in default resolver first.
+                        // If we are running in .NET core, we ask the built-in default resolver first.
```
Comment (JanKrivanek) [inline src/Build/BackEnd/Components/SdkResolution/SdkResolverService.cs:113]: nit: whitespace

File: src/Build/BackEnd/Components/SdkResolution/SdkResolverService.cs [hunk 3]
```diff
@@ -165,27 +157,8 @@ public virtual SdkResult ResolveSdk(int submissionId, SdkReference sdk, LoggingC
...
-                    {
-                        loggingContext.LogError(new BuildEventFileInfo(sdkReferenceLocation), "FailedToResolveSDK", sdk.Name, string.Join($"{Environment.NewLine}  ", errors));
-                    }
-
-                    LogWarnings(loggingContext, sdkReferenceLocation, warnings);
-                }
 
-                return result;
-            }
+            return ResolveSdkUsingResolversWithPatternsFirst(submissionId, sdk, loggingContext, sdkReferenceLocation, solutionPath, projectPath, interactive, isRunningInVisualStudio, failOnUnresolvedSdk);
```
Comment (JanKrivanek) [inline src/Build/BackEnd/Components/SdkResolution/SdkResolverService.cs:161]: Is this a functional change? Should it possibly be a separate PR or at least commit (if not squashing the PR)?

File: src/Build/Construction/ProjectRootElement.cs [hunk 4]
```diff
@@ -1691,31 +1691,13 @@ private void ReloadFrom(Func<bool, XmlDocumentWithLocation> documentProducer, bo
...
-                    oldDocument.ClearAnyCachedStrings();
-                }
-                if (!object.ReferenceEquals(currentDocument, newDocument))
-                {
-                    newDocument.ClearAnyCachedStrings();
-                }
-            }
+            RemoveAllChildren();
+
+            ProjectParser.Parse(newDocument, this);
```
Comment (JanKrivanek) [inline src/Build/Construction/ProjectRootElement.cs:1700]: Why is this change part of changewaves retiring?

**Review by JanKrivanek — COMMENTED** [review]
Overall looks good.
Before signing off it'd be nice to know more details about code being removed that's not part of changewaves condition on a first look (I know big portion of that is just port from the previous PR - but I didn't see the context there either)

File: src/Build/Construction/ProjectRootElement.cs (hunk 4 above)
Comment (Forgind) [reply src/Build/Construction/ProjectRootElement.cs:1700]: There was a change wave governing what the string cache did: if the change wave was enabled, it basically just returned the string without touching it or interning it. Since that was the point of the string cache, this change removed the whole class and all references to it, including those in this finally block.

File: src/Build/BackEnd/Components/SdkResolution/SdkResolverService.cs (hunk 3 above)
Comment (Forgind) [reply src/Build/BackEnd/Components/SdkResolution/SdkResolverService.cs:161]: I'm a little confused; this line was in the `if (ChangeWaves.AreFeaturesEnabled(ChangeWaves.Wave17_4))` block, and the deleted code was in the else, so this seems like a functional change only if the 17.4 change wave was disabled.

**Comment by rainersigwald** [comment]
> Also, I thought we were planning to always have three change waves active at a time? If we're changing that, perhaps we should update the .md to say that 🙂

I interpreted

> A wave of features is set to "rotate out" (i.e. become standard functionality) two bands after its release. For example, wave 16.8 stayed opt-out through wave 16.10, becoming standard functionality when wave 17.0 is introduced.

and

> Invalid Value (Ex: `16.9` when valid waves are `16.8` and `16.10`)

To mean we should generally only have two available but I couldn't remember the details on this part. @marcpopMSFT do you, by chance?

File: src/Build/ElementLocation/XmlDocumentWithLocation.cs [hunk 5]
```diff
@@ -262,7 +240,7 @@ public override XmlWhitespace CreateWhitespace(string text)
                 text = String.Empty;
             }
 
-            string interned = StringCache.Add(text, this);
+            string interned = Strings.WeakIntern(text);
```
[the remark this answers is not in the document]
Comment (rainersigwald) [reply src/Build/ElementLocation/XmlDocumentWithLocation.cs:243]: @AR-May this came from your changes, got an opinion?

File: src/Build/BackEnd/Components/SdkResolution/SdkResolverService.cs (hunk 3 above)
Comment (rainersigwald) [reply src/Build/BackEnd/Components/SdkResolution/SdkResolverService.cs:161]: @JanKrivanek that was indeed the intent, are you seeing something else?

**Comment by Forgind** [comment]
> > Also, I thought we were planning to always have three change waves active at a time? If we're changing that, perhaps we should update the .md to say that 🙂
> 
> I interpreted
> 
> > A wave of features is set to "rotate out" (i.e. become standard functionality) two bands after its release. For example, wave 16.8 stayed opt-out through wave 16.10, becoming standard functionality when wave 17.0 is introduced.
> 
> and
> 
> > Invalid Value (Ex: `16.9` when valid waves are `16.8` and `16.10`)
> 
> To mean we should generally only have two available but I couldn't remember the details on this part. @marcpopMSFT do you, by chance?

It seems a bit ambiguous. I saw examples like this:
https://github.com/dotnet/msbuild/blob/6fdd7318695eea68bf51bff5d6caf19f079a8597/documentation/wiki/ChangeWaves-Dev.md?plain=1#L38
which has three active change waves after the update. I interpreted `two bands after its release` to mean that we'd have 17_0 and 17_2 would be released (and active), and 17_4 would be active but not released (because it's in preview). Then when 17_4 releases, 17_0 would rotate out, and 17_6 would become the new preview version.

Regardless, I don't think it actually matters, so I have no problem with your interpretation and making that the official version.

**Comment by rainersigwald** [comment]
Ah; I think that makes sense in e.g. the 17.3 release--but now we're in the 17.12 release so that one will be "released and active", right?

File: src/Build/BackEnd/Components/SdkResolution/SdkResolverService.cs (hunk 3 above)
Comment (JanKrivanek) [reply src/Build/BackEnd/Components/SdkResolution/SdkResolverService.cs:161]: Sorry - my confusion only (I'm looking on the change and my comment and cannot recall what was I thinking)

**Review by JanKrivanek — APPROVED** [review]
Let's slash that code

File: src/Build/ElementLocation/XmlDocumentWithLocation.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (rainersigwald) [reply src/Build/ElementLocation/XmlDocumentWithLocation.cs:243]: I really want this in for 17.12-preview1 so I'm going to go in with the current state. https://github.com/dotnet/msbuild/issues/10465 for follow up.

**Comment by Forgind** [comment]
> Ah; I think that makes sense in e.g. the 17.3 release--but now we're in the 17.12 release so that one will be "released and active", right?

I'd say we're working on 17.12, so it's active but still in preview, that is, not released.
